### ISSUE TITLE
Plugin behaviour should be Opt-in, not Opt-out

### ISSUE THREAD

**Issue by sect2k** [chartjs/chartjs-plugin-datalabels#42]
First of all thanks for your work on this plugin, much appreciated.

Right now, when adding the script to page or importing it, it automatically add itself to all charts, which is quite a big assumption on how it's to be used and requires a lot of work to disable it on instance to instance basis.

**Comment by simonbrunel** [comment]
This has already been (more or less) discussed in #8 and #20. I *think* I agree that it should be disabled by default, however it's currently not that complicated to disable labels for all instances, it's one line:

```javascript
// Globally disable datalabels
Chart.defaults.global.plugins.datalabels.display = false
```

This default behavior can't be changed until the next major release since it's a breaking change. Though, I'm not sure about changing it because it may generate more tickets from people reporting that the plugin doesn't work because they didn't set `display: true`.

**Comment by sect2k** [comment]
@simonbrunel thanks for the quick reply, I ended up doing something similar. I guess I should have searched better, didn't occur to me to check closed issues.

A possible solution would be to keep current behavior for the UMD build, but have it be opt-in when imported as a module. This way most people would not be affected by this change.

**Comment by kumarharsh** [comment]
Okay, I figured out what was going wrong for solution 1. I had to use the **same instance** of the plugin as returned by `Chart.plugins.getAll()`! The way I had organized my react modules made me write this:

```
import { Chart } from 'react-chartjs-2';
import 'chartjs-plugin-datalabels';

const [dataLabelsPlugin] = Chart.plugins
  .getAll()
  .filter(p => p.id === 'datalabels');
Chart.plugins.unregister(dataLabelsPlugin);

/* @HACK: globally unregister the datalabels plugin because it can't be enabled
per instance currently. Remove this when library updates. */
export { default as annotationPlugin } from 'chartjs-plugin-annotation';
export { dataLabelsPlugin };
```

Just leaving this here if someone else stumbles upon this problem.

**Comment by simonbrunel** [comment]
> Given how the annotation plugin behaves ...

What do you mean? It seems to [registers itself automatically](/chartjs/chartjs-plugin-annotation/blob/master/src/index.js#L47) as well, the only difference is that it also exports itself, which makes easier to globally unregister it:

```javascript
import annotation from 'chartjs-plugin-annotation';

Chart.plugins.unregister(annotation);
```

I agree that in case of module imports, it would be better if the plugin doesn't register itself automatically (that's why I kept this ticket open and will change the behavior in 1.x). But first we need to figure out what to do in case the plugin is imported via HTML script? How to get the plugin handle to be able to register it locally or globally?

```html
<script src="path/to/chartjs/dist/chartjs-plugin-datalabels.js"></script>
<script>
    var plugin = // ????
</script>
```

**Comment by simonbrunel** [comment]
@kumarharsh Can you share references about how to implement your suggestion? (we use rollup)

I found that rollup defines a global variable with the default export, so we could simply remove the auto-registration for everyone and do the following when using `<script>`:

```html
<script src="path/to/chartjs-plugin-datalabels.js"></script>
<script>
    // Window.ChartJsPluginDatalabels is exported by rollup 

    // register globally
    Chart.plugins.register(ChartJsPluginDatalabels);

    // register locally
    new Chart('id', {
        plugins: [ChartJsPluginDatalabels]
    });
</script>
```

However, when using `<script>`, I'm not sure users want that extra step in order to use this plugin, but I could be wrong, in which case it makes things easier :)

**Comment by kumarharsh** [comment]
I'll look into rollup and see how it can be done, probably over the weekend... In webpack, I use the `webpack.DefinePlugin` to do this.

**Comment by simonbrunel** [comment]
Thanks @kumarharsh (please share your code here instead of creating a PR since these changes will not happen before v1).

I would actually prefer the same behavior for everyone (no auto-register) but since I'm not using Chart.js (and so plugins), I don't know if my previous suggestion is acceptable for someone importing via `<script>`. @sect2k @jledentu @kumarharsh what do you think?

**Comment by jledentu** [comment]
@simonbrunel I agree with you that it's better to have the same behaviour for everyone.

Some popular libraries have a different behaviour between AMD/CommonJS imports and includes as script. In Vue for instance, some plugins like the popular `vue-router` is used automatically if it's available as global variable, whereas it must be used explicitly when imported as AMD/CommonJS module. I personnaly don't like this behaviour that can create confusion, and inconsistency in docs/examples.

So in my opinion, your suggestion is totally acceptable. 👍

**Comment by kumarharsh** [comment]
I guess you can name it `window.ChartjsDataLabels` to scope it better. But apart from that, the suggestion seems solid. I personally would opt for the approach of keeping it explicit everywhere and adding a line in docs about usage.

**Comment by simonbrunel** [comment]
That was my first idea but since the Chart.js class is called `Chart`, I think it's more consistent to call it `ChartDataLabels` (like `Vue` and `VueRouter`, not `VuejsRouter`).

>  I personally would opt for the approach of keeping it explicit everywhere...

I agree, starting v1.0, this plugin will not be automatically registered whatever the import method.

**Comment by simonbrunel** [comment]
A first step implemented in 957ee999db (default export but still auto-registered) and [documentation](https://master--chartjs-plugin-datalabels.netlify.com/guide/getting-started.html#registration) updated about the current behavior until version 1.

**Comment by benmccann** [comment]
If we do not auto-register, is it possible to register a plugin only for a single chart? It seems we might need to add the functionality to the main Chart.js library or better document it if it already exists. It looks to me like the `register` function can only be called globally today: https://www.chartjs.org/docs/latest/developers/plugins.html

**Comment by simonbrunel** [comment]
@sect2k @jledentu @kumarharsh I'm looking to (finally) release v1 and I would like to check with you if removing the auto-registration entirely (`<script>` and `import`) is still the way to go. I think that a consistent behavior is better but some other official plugins are taking different paths. For example, the `annotation` registers the plugin in UMD builds but not in ESM, which seems even more confusing since UMD can also be used with `require/import` - or I misunderstood something, which may actually be the case :) Ideally, we should also have consistency in the way official plugins behave.

What would be the benefits of auto-registering the plugin in some cases but not in others?

**Comment by simonbrunel** [comment]
Done in 13daa478d9082fd0bc100e36394719215771d689 and will be released in v1.0.0. [Getting Started](https://master--chartjs-plugin-datalabels.netlify.app/guide/getting-started.html#registration) updated and added this breaking change to the [migration guide](https://master--chartjs-plugin-datalabels.netlify.app/guide/migration.html).

**Comment by s4m0r4m4** [comment]
Hi @simonbrunel, I just started using your plugin (0.7.0) and came across the auto-register behavior. Just one person's perspective here, but I do think the new direction you're going with in v1.0 makes more sense to me. Thanks again for an amazing plugin!

**Comment by simonbrunel** [comment]
Closing since it has been released in [v1.0.0-beta.1](https://github.com/chartjs/chartjs-plugin-datalabels/releases/tag/v1.0.0-beta.1) (Chart.js 2.x) and [v2.0.0-beta.1](https://github.com/chartjs/chartjs-plugin-datalabels/releases/tag/v2.0.0-beta.1) (Chart.js 3.x) 

Any help testing these 2 beta versions is welcome :)

Thanks everyone for your feedback.
